app/controllers/name_controller/create_and_edit_name.rb

```diff
@@ -223,73 +229,131 @@ def parse_name
+      @name.real_text_name
+    else
+      params[:name][:text_name]
     end
   end
 
-  def check_for_matches
-    matches = Name.where(search_name: @parse.search_name) - [@name]
-    return matches.first unless matches.many?
-
-    args = {
-      str: @parse.real_search_name,
-      matches: new_name.map(&:search_name).join(" / ")
-    }
-    raise(:edit_name_multiple_names_match.t(args))
-  end
-
-  def status_changing?
-    params[:name][:deprecated].to_s != @name.deprecated.to_s
-  end
+  def set_name_author_and_rank
+    return unless name_unlocked?
 
-  def name_unlocked?
-    in_admin_mode? || !@name.locked
+    email_admin_name_change unless minor_change?
+    @name.attributes = @parse.params
   end
 
   def minor_change?
+    return false if icn_id_conflict?(params[:name][:icn_id])
+    return true if just_adding_author?
+
     old_name = @name.real_search_name
     new_name = @parse.real_search_name
     new_name.percent_match(old_name) > 0.9
   end
 
+  def icn_id_conflict?(new_icn_id)
+    new_icn_id && @name.icn_id &&
+      new_icn_id != @name.icn_id
+  end
+
   def just_adding_author?
     @name.author.blank? && @parse.text_name == @name.text_name
   end
 
+  def update_ancestors
+    Name.find_or_create_parsed_name_and_parents(@parse).each do |name|
+      name.save_with_log(:log_name_created) if name&.new_record?
+    end
+  end
+
   def email_admin_name_change
     subject = "Nontrivial Name Change"
-    content = :email_name_change.l(
-      user: @user.login,
+    content = :email_name_change.l(email_name_change_content)
+    WebmasterEmail.build(@user.email, content, subject).deliver_now
+    NameControllerTest.report_email(content) if Rails.env.test?
+  end
+
+  def email_name_change_content
+    { user: @user.login,
+      old_identifier: @name.icn_id,
+      new_identifier: params[:name][:icn_id],
       old: @name.real_search_name,
       new: @parse.real_search_name,
       observations: @name.observations.length,
       namings: @name.namings.length,
-      url: "#{MO.http_domain}/name/show_name/#{@name.id}"
-    )
-    WebmasterEmail.build(@user.email, content, subject).deliver_now
-    NameControllerTest.report_email(content) if Rails.env.test?
+      url: "#{MO.http_domain}/name/show_name/#{@name.id}" }
   end
 
-  def redirect_to_show_name
-    redirect_with_query(@name.show_link_args)
-  end
+  # -------------
+  # update: merge
+  # -------------
 
-  def redirect_to_approve_or_deprecate
-    if params[:name][:deprecated].to_s == "true"
-      redirect_with_query(action: :deprecate_name, id: @name.id)
+  def merge_names(new_name)
+    if in_admin_mode? || @name.mergeable? || new_name.mergeable?
+      perform_merge_names(new_name)
+      redirect_to_show_name
     else
-      redirect_with_query(action: :approve_name, id: @name.id)
+      redirect_to_merge_request(new_name)
+    end
+  end
+
+  def perform_merge_names(new_name)
+    old_display_name_for_log = @name[:display_name]
```

Comment:
I don't understand the 3 lines involving @name[:display_name] and @name.display_name: this line, line 315, and line 317 --
- Why is the attribute being accessed via a hash key here and in line 315?
- Why do we need these lines at all?

Reply:
Check out app/models/name/format.rb.  Apparently we override display_name.